api-compose-object.go

```diff
@@ -0,0 +1,392 @@
+type SSEInfo struct {
+	key  []byte
+	algo string
+}
+
+// NewSSEInfo - specifies (binary or un-encoded) encryption key and
+// algorithm name. If algo is empty, it defaults to "AES256".
+func NewSSEInfo(key []byte, algo string) SSEInfo {
+	if algo == "" {
+		algo = "AES256"
+	}
+	return SSEInfo{key, algo}
+}
+
+// internal method to output the generated SSE headers.
+func (e *SSEInfo) getSSEHeaders(isCopySource bool) map[string]string {
+	if e == nil {
+		return nil
+	}
+
+	cs := ""
+	if isCopySource {
+		cs = "copy-source-"
+	}
+	return map[string]string{
+		"x-amz-" + cs + "server-side-encryption-customer-algorithm": e.algo,
+		"x-amz-" + cs + "server-side-encryption-customer-key":       base64.StdEncoding.EncodeToString(e.key),
+		"x-amz-" + cs + "server-side-encryption-customer-key-MD5":   base64.StdEncoding.EncodeToString(sumMD5(e.key)),
+	}
+}
+
+// DestinationInfo - type with information about the object to be
+// created via server-side copy requests, using the Compose API.
+type DestinationInfo struct {
+	bucket, object string
+	encryption     *SSEInfo
+	customMetadata map[string]string
+}
+
+// NewDestinationInfo - creates a DesinationInfo instance -
+// encryptionInfo may be nil to indicate that no-encryption is to be
+// done, and metadata may be nil to indicate that source metadata is
+// to be copied to the destination object from the source.
+func NewDestinationInfo(bucket, object string, encryptionInfo *SSEInfo,
+	metadata map[string]string) DestinationInfo {
+
+	return DestinationInfo{
+		bucket:         bucket,
+		object:         object,
+		encryption:     encryptionInfo,
+		customMetadata: metadata,
+	}
+}
+
+// getCustomHeadersMap - construct appropriate key-value pairs to send
+// as headers from metadata map to pass into copy-object request. For
+// single part copy-object (i.e. non-multipart object), enable the
+// withCopyDirectiveHeader to set the `x-amz-metadata-directive` to
+// `REPLACE`, so that metadata headers from the source are not copied
+// over.
+func (d *DestinationInfo) getCustomHeadersMap(withCopyDirectiveHeader bool) map[string]string {
+	if d == nil || d.customMetadata == nil {
+		return nil
+	}
+	r := make(map[string]string)
+	if withCopyDirectiveHeader {
+		r["x-amz-metadata-directive"] = "REPLACE"
+	}
+	for k, v := range d.customMetadata {
+		if !strings.HasPrefix(k, "x-amz-meta-") {
+			k = "x-amz-meta-" + k
+		}
+		r[k] = v
+	}
+	return r
+}
+
+// SourceInfo - represents a source object to be copied, using
+// server-side copying APIs.
+type SourceInfo struct {
+	bucket, object string
+
+	start, end int64
+
+	// Headers to send with the upload-part-copy request involving
+	// this source object.
+	Headers http.Header
+}
+
+// NewSourceInfo - create a new SourceInfo object. decryptionInfo may
+// be nil if the source object is not encrypted.
+func NewSourceInfo(bucket, object string, decryptionInfo *SSEInfo) SourceInfo {
+	r := SourceInfo{
+		bucket:  bucket,
+		object:  object,
+		start:   -1,
```

Comment:
I know start = -1 means start is unspecified but we can also set it to zero by default to simplify the code. Do we really need -1 state ?